src/persist/src/indexed/columnar/arrow.rs

```diff
@@ -117,6 +98,74 @@ pub(crate) const ENABLE_ARROW_LGALLOC_NONCC_SIZES: Config<bool> = Config::new(
     "A feature flag to enable copying decoded arrow data into lgalloc on non-cc sized clusters.",
 );
 
+fn realloc_data(data: ArrayData, metrics: &ColumnarMetrics) -> ArrayData {
+    // NB: Arrow generally aligns buffers very coarsely: see arrow::alloc::ALIGNMENT.
+    // However, lgalloc aligns buffers even more coarsely - to the page boundary -
+    // so we never expect alignment issues in practice. If that changes, build()
+    // will return an error below, as it does for all invalid data.
+    let buffers = data
+        .buffers()
+        .iter()
+        .map(|b| realloc_buffer(b, metrics))
+        .collect();
+    let child_data = data
+        .child_data()
+        .iter()
+        .cloned()
+        .map(|d| realloc_data(d, metrics))
+        .collect();
+    let nulls = data.nulls().map(|n| {
+        let buffer = realloc_buffer(n.buffer(), metrics);
+        NullBuffer::new(BooleanBuffer::new(buffer, n.offset(), n.len()))
+    });
+
+    data.into_builder()
+        .buffers(buffers)
+        .child_data(child_data)
+        .nulls(nulls)
+        .build()
+        .expect("reconstructing valid arrow array")
+}
+
+/// Re-allocate the backing storage for a specific array using lgalloc, if it's configured.
+pub fn realloc_array<A: Array + From<ArrayData>>(array: &A, metrics: &ColumnarMetrics) -> A {
+    let data = array.to_data();
+    let data = realloc_data(data, metrics);
+    A::from(data)
+}
+
+/// Re-allocate the backing storage for an array ref using lgalloc, if it's configured.
+pub fn realloc_any(array: &ArrayRef, metrics: &ColumnarMetrics) -> ArrayRef {
+    let data = array.to_data();
+    let data = realloc_data(data, metrics);
+    make_array(data)
+}
+
+fn realloc_buffer(buffer: &Buffer, metrics: &ColumnarMetrics) -> Buffer {
+    let use_lgbytes_mmap = if metrics.is_cc_active {
+        ENABLE_ARROW_LGALLOC_CC_SIZES.get(&metrics.cfg)
+    } else {
+        ENABLE_ARROW_LGALLOC_NONCC_SIZES.get(&metrics.cfg)
+    };
+    let region = if use_lgbytes_mmap {
+        metrics
+            .lgbytes_arrow
+            .try_mmap_region(buffer.as_slice())
+            .ok()
+    } else {
+        None
+    };
+    let Some(region) = region else {
+        return buffer.clone();
+    };
+    let bytes: &[u8] = region.as_ref().to_byte_slice();
+    let ptr: NonNull<[u8]> = bytes.into();
+    // This is fine: see [[NonNull::as_non_null_ptr]] for an unstable version of this usage.
+    let ptr: NonNull<u8> = ptr.cast();
+    // SAFETY: `ptr` is valid for `len` bytes, and kept alive as long as `region` lives.
+    unsafe { Buffer::from_custom_allocation(ptr, bytes.len(), Arc::new(region)) }
```

Comment:
Fairly confident I've upheld the safety requirements for this API, but I'd appreciate extra scrutiny on it!

Reply:
From the [`arrow-rs` docs](https://docs.rs/arrow/latest/arrow/buffer/struct.Buffer.html#method.from_custom_allocation):
> This function is unsafe as there is no guarantee that the given pointer is valid for len bytes

Which it looks like we're doing here!